modules/dag-l1/src/main/scala/org/tessellation/dag/l1/infrastructure/block/BlockService.scala

```diff
@@ -0,0 +1,173 @@
+          _ <- updateTipParentsToRemain(tipParentsToRemain.keySet.toList, hashedBlock)
+          _ <- tryAddAsATip(addAsTipCondition, hashedBlock)
+        } yield ()
+
+      def pullTips(tipsCount: PosInt): F[Option[Tips]] =
+        for {
+          allBlocks <- blocks.toMap
+          allTips = allBlocks.collect { case (_, AcceptedBlock(block, _, _, true)) => block }.toList
+            .sortBy(_.signed.value.height.coerce)
+          pulledTips = allTips match {
+            case tips if tips.size >= tipsCount =>
+              NonEmptyList
+                .fromList(tips.take(tipsCount))
+                .map(parents => block.Tips(parents.map(p => BlockReference(p.proofsHash, p.signed.value.height))))
+            case _ => None
+          }
+        } yield pulledTips
+
+      private def isBlockAccepted(blockReference: BlockReference): F[Boolean] =
+        blocks(blockReference.hash).get.map(_.exists(_.isInstanceOf[AcceptedBlock]))
+
+      private def updateNoTipParents(parents: List[ProofsHash], hashedBlock: Hashed[DAGBlock]): F[Unit] =
+        parents.traverse { hash =>
+          blocks(hash)
+            .modify[Either[TipsUpdateError, Unit]] {
+              case Some(accepted @ AcceptedBlock(_, _, _, false)) =>
+                (accepted.addUsage(hashedBlock).some, ().asRight)
+              case unexpected =>
+                (unexpected, NoTipAcceptedBlockNotFound(hash, unexpected).asLeft)
+            }
+            .flatMap(_.liftTo[F])
+        }.void
+
+      private def updateTipParentsToRemove(parents: List[ProofsHash], hashedBlock: Hashed[DAGBlock]): F[Unit] =
+        parents.traverse { hash =>
+          blocks(hash)
+            .modify[Either[TipsUpdateError, Unit]] {
+              case Some(accepted @ AcceptedBlock(_, _, _, true)) =>
+                (accepted.addUsage(hashedBlock).unsetAsTip.some, ().asRight)
+              case unexpected =>
+                (unexpected, TipBlockToUnsetAsTipNotFound(hash, unexpected).asLeft)
+            }
+            .flatMap(_.liftTo[F])
+        }.void
+
+      private def updateTipParentsToRemain(parents: List[ProofsHash], hashedBlock: Hashed[DAGBlock]): F[Unit] =
+        parents.traverse { hash =>
+          blocks(hash)
+            .modify[Either[TipsUpdateError, Unit]] {
+              case Some(accepted @ AcceptedBlock(_, _, _, true)) =>
+                (accepted.addUsage(hashedBlock).some, ().asRight)
+              case unexpected =>
+                (unexpected, TipBlockToRemainAsTipNotFound(hash, unexpected).asLeft)
+            }
+            .flatMap(_.liftTo[F])
+        }.void
+
+      private def tryAddAsATip(condition: Boolean, hashedBlock: Hashed[DAGBlock]): F[Unit] =
+        if (condition)
+          blocks(hashedBlock.proofsHash).modify {
+            case Some(accepted @ AcceptedBlock(_, _, usages, false))
+                if usages.size <= tipsConfig.maximumTipUsages - 1 =>
```

Comment:
I see the same condition twice

Reply:
I will check if somehow extracting it for both cases makes sense.

Reply:
This one ones simply needs to be `< maximumTipUsages` don't know why I did `<=` and subtracted `1`.